core/src/commonMain/kotlin/SharedRepeatableAction.kt

```diff
@@ -46,34 +48,50 @@ internal class SharedRepeatableAction<T>(
     suspend fun await() = getOrAsync().await()
 
     @Suppress("ktlint:standard:indent")
-    internal fun getOrAsync(): Deferred<T> = guard.withLock {
+    private fun getOrAsync(): Deferred<T> = guard.withLock {
         (
             state?.takeUnless { it.isCancelled } ?: run {
                 val rootJob = Job(coroutineContext.job)
                 val rootScope = CoroutineScope(coroutineContext + rootJob)
-                val actionJob = rootScope.async { action(rootScope) }.apply {
+                val actionJob = rootScope.async {
+                    try {
+                        action(rootScope)
+                    } catch (e: CancellationException) {
+                        (e.cause as? SharedRepeatableActionException)?.cause?.let { throw it }
+                        throw e
+                    }
+                }.apply {
                     invokeOnCompletion { cause ->
-                        if (cause != null) this@SharedRepeatableAction.reset()
+                        if (cause != null) setCancelled()
                     }
                 }
                 State(rootJob, actionJob)
             }.also { state = it }
         ).action
     }
 
-    fun reset() {
-        guard.withLock {
-            state?.apply { isCancelled = true }
-        }?.root?.cancel()
+    fun fail(cause: Throwable) {
+        setCancelled()
+            ?.root
+            ?.cancel(CancellationException(null, SharedRepeatableActionException(cause)))
```

Comment:
It would've been much cleaner to have `SharedRepeatableActionException` extend `CancellationException` and cancel with `SharedRepeatableActionException` directly, but I couldn't figure out how to extend `CancellationException`. 😢 

So, we have to do the dance on line 60, where we check the type of the `cause`.